tket/src/Architecture/Architecture.cpp

```diff
@@ -25,33 +25,78 @@
 
 namespace tket {
 
-// basic implementation that works off same prior assumptions
-// TODO: Update this for more mature systems of multi-qubit gates
 bool Architecture::valid_operation(
     const Op_ptr& op, const std::vector<Node>& uids) const {
-  if (op->get_desc().is_box() ||
-      (op->get_type() == OpType::Conditional &&
-       static_cast<const Conditional&>(*op).get_op()->get_desc().is_box())) {
+  // boxes are never allowed
+  if (is_box_type(op->get_type())) {
     return false;
   }
-  if (uids.size() == 1) {
-    // with current Architecture can assume all single qubit gates valid
-    return true;
-  } else if (op->get_type() == OpType::Barrier) {
+
+  // TODO
+  // this currently allows unplaced single qubits gates
+  // this should be changes in the future
+  if (is_single_qubit_type(op->get_type())) {
     return true;
-  } else if (uids.size() == 2) {
-    if (this->node_exists(uids[0]) && this->node_exists(uids[1]) &&
-        this->bidirectional_edge_exists(uids[0], uids[1])) {
-      return true;
+  }
+
+  if (op->get_type() == OpType::Conditional) {
+    auto ot = static_cast<const Conditional&>(*op).get_op()->get_type();
+    // conditional boxes are never allowed, too
+    if (is_box_type(ot)) {
+      return false;
     }
-  } else if (uids.size() == 3 && op->get_type() == OpType::BRIDGE) {
-    bool con_0_exists = this->bidirectional_edge_exists(uids[0], uids[1]);
-    bool con_1_exists = this->bidirectional_edge_exists(uids[2], uids[1]);
-    if (this->node_exists(uids[0]) && this->node_exists(uids[1]) &&
-        this->node_exists(uids[2]) && con_0_exists && con_1_exists) {
+
+    // TODO
+    // this currently allows unplaced single qubits gates
+    // this should be changes in the future
+    if (is_single_qubit_type(ot)) {
       return true;
     }
+    if (is_multi_qubit_type(ot)) {
+      if (uids.size() == 2) {
+        if (this->node_exists(uids[0]) && this->node_exists(uids[1]) &&
+            this->bidirectional_edge_exists(uids[0], uids[1])) {
+          return true;
+        }
+      } else if (uids.size() == 3 && ot == OpType::BRIDGE) {
+        bool con_0_exists = this->bidirectional_edge_exists(uids[0], uids[1]);
+        bool con_1_exists = this->bidirectional_edge_exists(uids[2], uids[1]);
+        if (this->node_exists(uids[0]) && this->node_exists(uids[1]) &&
+            this->node_exists(uids[2]) && con_0_exists && con_1_exists) {
+          return true;
+        }
+      }
+    }
   }
+
+  // Barriers are allways allowed
+  if (op->get_type() == OpType::Barrier) {
+    return true;
+  }
+
+  // allow two qubit gates only for placed and connected nodes
+  if (is_multi_qubit_type(op->get_type())) {
```

Comment:
It looks like you have some repeated code? I'm guessing a copy and paste gone wrong

Reply:
> To address your comment here - I like the safe checking of the OpTypes and whether they are single_qubit/multi_qubit here. However i'm wondering whether we've cursed ourselves with this name and are trying to over fit the solution, in a scenario where the performance hampering could be too large (as you point out).
> 
> Sorry if this undoes work. The largest problem here is that due to its name and being an Architecture method, the correct approach to writing this method (which I think is what you have done) is too much for Mapping. Hence I now suggest we do the following two things:
> 
> 1. Make Make a new MappingFrontier method, with similar name (valid_boundary_operation or something). Retain some of the same logic with these differences: a) return true if uids.size() == 1 always (remove single_qubit_type check), b) r emove multi_qubit_type check, but keep  how uids.size() ==2  and uids.size() == 3 checks works
> 2. Keep this Architecture::valid_operation as is and expose it to the pytket layer as an Architecture method. It will be useful for anybody making their own python level heuristics. Add python tests.
> 
> Does this sound reasonable to you?

Yes that sound reasonable. As allready mention I was not really happy with the longer runtime, the function is used very often so it could have a big impact at the end increasing the runtime only a bit.

I think it would be much more consistent to remove the box and conditional checks from the architecture valid command in there, but that would require major changes.

I will continue as suggested.

Reply:
One of the cases is handling the case of an conditional gate, so it is not fully identical